src/builder/Actions.sol

```diff
@@ -163,6 +185,71 @@ library Actions {
         return (quarkOperation, action);
     }
 
+    function cometSupplyAsset(CometSupply memory cometSupply, PaymentInfo.Payment memory payment)
+        internal
+        pure
+        returns (IQuarkWallet.QuarkOperation memory, Action memory)
+    {
+        bytes[] memory scriptSources = new bytes[](1);
+        scriptSources[0] = type(CometSupplyActions).creationCode;
+
+        Accounts.ChainAccounts memory accounts =
+            Accounts.findChainAccounts(cometSupply.chainId, cometSupply.chainAccountsList);
+
+        Accounts.AssetPositions memory assetPositions =
+            Accounts.findAssetPositions(cometSupply.assetSymbol, accounts.assetPositionsList);
+
+        Accounts.QuarkState memory accountState = Accounts.findQuarkState(cometSupply.sender, accounts.quarkStates);
+
+        bytes memory scriptCalldata;
+        if (Strings.stringEqIgnoreCase(cometSupply.assetSymbol, "ETH")) {
+            // XXX handle wrapping ETH
+        } else {
+            scriptCalldata = abi.encodeWithSelector(
+                CometSupplyActions.supply.selector, cometSupply.comet, assetPositions.asset, cometSupply.amount
+            );
+        }
+        // Construct QuarkOperation
+        IQuarkWallet.QuarkOperation memory quarkOperation = IQuarkWallet.QuarkOperation({
+            nonce: accountState.quarkNextNonce,
+            scriptAddress: CodeJarHelper.getCodeAddress(type(CometSupplyActions).creationCode),
+            scriptCalldata: scriptCalldata,
+            scriptSources: scriptSources,
+            expiry: cometSupply.blockTimestamp + STANDARD_EXPIRY_BUFFER
+        });
+
+        if (payment.isToken) {
+            // Wrap operation with paycall
+            quarkOperation = PaycallWrapper.wrap(
+                quarkOperation,
+                cometSupply.chainId,
+                payment.currency,
+                PaymentInfo.findMaxCost(payment, cometSupply.chainId)
+            );
+        }
```

Comment:
Probably need to handle supplyMax case as well. 
But since I've done transferMax, I can also help change this in a follow-up PR if wanted. 

Maybe can add a `TODO:` here, so it's easier to find later one 😄

Reply:
added a todo in QuarkBuilder.sol